library/core/src/ptr/mut_ptr.rs

```diff
@@ -1338,6 +1338,111 @@ impl<T> *mut [T] {
+    /// the index `mid` itself) and the second will contain all
+    /// indices from `[mid, len)` (excluding the index `len` itself).
+    ///
+    /// # Panics
+    ///
+    /// Panics if `mid > len`.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// #![feature(raw_slice_split)]
+    /// #![feature(slice_ptr_get)]
+    ///
+    /// let mut v = [1, 0, 3, 0, 5, 6];
+    /// let ptr = &mut v as *mut [_];
+    /// let (left, right) = ptr.split_at_mut(2);
+    /// unsafe {
+    ///     assert_eq!(&*left, [1, 0]);
+    ///     assert_eq!(&*right, [3, 0, 5, 6]);
+    /// }
+    /// ```
+    #[inline(always)]
+    #[track_caller]
+    #[unstable(feature = "raw_slice_split", issue = "95595")]
+    pub fn split_at_mut(self, mid: usize) -> (*mut [T], *mut [T]) {
+        assert!(mid <= self.len());
+        // SAFETY: `[ptr; mid]` and `[mid; len]` are inside `self`, which
+        // fulfills the requirements of `from_raw_parts_mut`.
+        unsafe { self.split_at_mut_unchecked(mid) }
+    }
+
+    /// Divides one mutable raw slice into two at an index, without doing bounds checking.
+    ///
+    /// The first will contain all indices from `[0, mid)` (excluding
+    /// the index `mid` itself) and the second will contain all
+    /// indices from `[mid, len)` (excluding the index `len` itself).
+    ///
+    /// For a safe alternative see [`split_at_mut`].
+    ///
+    /// [`split_at_mut`]: #method.split_at_mut
+    ///
+    /// # Safety
+    ///
+    /// Calling this method with an out-of-bounds index is *[undefined behavior]*
+    /// even if the resulting reference is not used. The caller has to ensure that
+    /// `0 <= mid <= self.len()`.
+    ///
+    /// [undefined behavior]: https://doc.rust-lang.org/reference/behavior-considered-undefined.html
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// #![feature(raw_slice_split)]
+    ///
+    /// let mut v = [1, 0, 3, 0, 5, 6];
+    /// // scoped to restrict the lifetime of the borrows
+    /// unsafe {
+    ///     let ptr = &mut v as *mut [_];
+    ///     let (left, right) = ptr.split_at_mut_unchecked(2);
+    ///     assert_eq!(&*left, [1, 0]);
+    ///     assert_eq!(&*right, [3, 0, 5, 6]);
+    ///     (&mut *left)[1] = 2;
+    ///     (&mut *right)[1] = 4;
+    /// }
+    /// assert_eq!(v, [1, 2, 3, 4, 5, 6]);
+    /// ```
+    #[inline(always)]
+    #[unstable(feature = "raw_slice_split", issue = "95595")]
+    pub unsafe fn split_at_mut_unchecked(self, mid: usize) -> (*mut [T], *mut [T]) {
+        let len = self.len();
+        let ptr = self.as_mut_ptr();
+
+        // SAFETY: Caller has to check that `0 <= mid <= self.len()`.
+        //
+        // `[ptr; mid]` and `[mid; len]` are not overlapping, so returning a mutable reference
+        // is fine.
+        unsafe {
+            (
+                crate::ptr::slice_from_raw_parts_mut(ptr, mid),
+                crate::ptr::slice_from_raw_parts_mut(ptr.add(mid), len - mid),
```

Comment:
`ptr.add` here is wrong I think since it might not actually be inbounds?
```rust
let slice_ref: *mut [u8] = &mut [0, 0, 0, 0];
let slice_ref: *mut [u128] = slice_ref as *mut [u128];
let (_, oops_not_inbounds) = slice_ref.split_at_mut(2);
```
the `// SAFETY` above this unsafe block doesn't talk about `ptr::add` either

Reply:
Isn't that covered by the API obligations for the caller?

> /// # Safety
> /// Calling this method with an out-of-bounds index is *[undefined behavior]*

Reply:
the `out-of-bounds` there is referring to the slice len i think, not "inboudns" as in the actual underlying allocation that pointer is pointing to. for `&mut [T]` thats the same thing, but for `*mut [T]` they are distinct concepts (as seen by the code i posted which has a `*mut [u128; 4]` pointing at an allocation of `*mut [u8; 4]`)

Reply:
The existing `get_unchecked_mut` phrases it as follows and does mean inbounds, since it also uses `add` internally

>    /// Calling this method with an out-of-bounds index or when `self` is not dereferenceable
>    /// is *[undefined behavior]* even if the resulting pointer is not used.

But you're right, it could be made clearer and the new methods are copy-pasted from `&mut [T]`, so they're still referring to references.

Reply:
Thinking about it, that also means that a `split_at_mut` must either be unsafe too or use `wrapping_add` since we wouldn't know whether it's applied to a valid pointer.

I think I'll make it unsafe so we get the bounds checks but the caller still has to use a valid pointer. But that means #94247 will need more unsafe blocks. 😔

Reply:
I don't really know what `get_unchecked_mut` has to do with it :sweat_smile:  The safety comments explicitly clarify that "out of bounds" is referrign to the slice length as it gives the example of: `0 <= mid <= self.len()` having to hold. The problem is that for the `&mut [T]` that invariant implies that you can offset the ptr and it be inbounds the allocation, but for `*mut [T]` that does not hold.

The `SAFETY` on `&mut [T]::split_at_mut_unchecked` imo could be considered somewhat incomplete reasoning wise. it just states `` // SAFETY: Caller has to check that `0 <= mid <= self.len()`.`` but the complete reasoning here includes the fact that `&mut [T]` means that the ptr points to an allocation of size `self.len() * size_of::<T>()` therefore offseting the ptr by `mid` must be inbounds.

with `*mut [T]::split_at_mut_unchecked` that "unspoken" logic is broken and so either
- we must not rely on it inside `*mut [T]::split_at_mut_unchecked` i.e. use `.wrapping_add`
OR
- we must add the invariant to the safety comments i.e. "the pointer must point to an allocation of atleast size `mid * size_of::<T>()`. Then if we look at `*mut [T]::split_at_mut` we can see that the invariant is not fulfilled (and then we can fix that by removing the method or making it unsafe)

Reply:
The latest commit hopefully addresses those concerns.